### DISCUSSION TITLE
Can't build Zola because of a relative path import

### DISCUSSION THREAD

**Discussion by bemyak** [ipetkov/crane#666] [Q&A] [answered]
Hi!

I maintain a tiny TUI page generator for Zola (a static site generator): https://gitlab.com/bemyak/zola-new

Since Zola itself is not published on crates.io, I have to import it as a git dependency:
```toml
[dependencies]
zola_site  = { git = "https://github.com/getzola/zola", tag = "v0.19.1", package = "site" }
zola_utils = { git = "https://github.com/getzola/zola", tag = "v0.19.1", package = "utils" }
```

However, building it fails because it uses git submodules in the root of the git repo:
```
error: builder for '/nix/store/0vv0vcfs1rny0sxwz10cyxh54zmm9hqs-zola-new-deps-0.1.5.drv' failed with exit code 101;
       last 10 log lines:
       >
       > error: couldn't read `/nix/store/falk5iwxsckx97rb39s9x0gsrsfkgkms-vendor-cargo-deps/590850d2007705cfc8f7b8b9a199fbfaee2b468f47dd897225d7f1a19398822d/config-0.1.0/src/../../../sublime/themes/all.themedump`: No such file or directory (os error 2)
       >   --> /nix/store/falk5iwxsckx97rb39s9x0gsrsfkgkms-vendor-cargo-deps/590850d2007705cfc8f7b8b9a199fbfaee2b468f47dd897225d7f1a19398822d/config-0.1.0/src/highlighting.rs:15:30
       >    |
       > 15 |     Lazy::new(|| from_binary(include_bytes!("../../../sublime/themes/all.themedump")));
       >    |                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
       >    |
       >    = note: this error originates in the macro `include_bytes` (in Nightly builds, run with -Z macro-backtrace for more info)
       >
       > error: could not compile `config` (lib) due to 2 previous errors
       For full logs, run 'nix log /nix/store/0vv0vcfs1rny0sxwz10cyxh54zmm9hqs-zola-new-deps-0.1.5.drv'
```

I've made sure to include those files in the src filter:
```nix
craneFilter = path: type: (
    builtins.match ".*\.packdump$" path != null ||
    builtins.match ".*\.themedump$" path != null ||
    craneLib.filterCargoSources path type
);
```

… but it didn't help :(

Could you please give me guidance on how to solve this issue?
Any help is greatly appreciated :pray:

**Comment by ipetkov** [comment, marked as answer]
Hi @bemyak! Typically crates should not attempt to read any files outside of crate's own "root" directory (i.e. where `Cargo.toml` file is stored). Git dependencies _happen_ to work (outside of Nix/crane) because of how cargo checks out the repo, but generally the same code would be broken if the code was ever published to a registry like crates.io.

You have two (well, three) options to resolve the issue:
1. Use your own fork which changes where the files are stored/included
2. Submit a fix upstream and wait for the project to accept it
3. Write some extra Nix code to patch the crate's sources during vendoring: https://crane.dev/patching_dependency_sources.html

Hope this helps!

**Reply by bemyak** [reply]
Thanks for the answer! I see where the issue is now :blush: 
I'll try to implement option 2, and if it's rejected I'll go with 3.
